cmd/ido2icingadb/convert.go

```diff
@@ -0,0 +1,827 @@
+			CancelTime:       cancelTime,
+			EndTime:          endTime,
+		}
+
+		s.DowntimeEnd.History = s
+		sla = append(sla, s)
+
+		checkpoint = row.DowntimehistoryId
+	}
+
+	icingaDbInserts = [][]interface{}{downtimeHistory, allHistory, sla}
+	return
+}
+
+// FlappingEnd updates an already migrated start event with the end event info.
+type FlappingEnd struct {
+	Id                    icingadbTypes.Binary    `json:"id"`
+	EndTime               icingadbTypes.UnixMilli `json:"end_time"`
+	PercentStateChangeEnd icingadbTypes.Float     `json:"percent_state_change_end"`
+}
+
+// Assert interface compliance.
+var _ contracts.TableNamer = (*FlappingEnd)(nil)
+
+// TableName implements the contracts.TableNamer interface.
+func (*FlappingEnd) TableName() string {
+	return "flapping_history"
+}
+
+type flappingRow = struct {
+	FlappinghistoryId  uint64
+	EventTime          int64
+	EventTimeUsec      uint32
+	EventType          uint16
+	PercentStateChange sql.NullFloat64
+	LowThreshold       float64
+	HighThreshold      float64
+	ObjecttypeId       uint8
+	Name1              string
+	Name2              string
+}
+
+func convertFlappingRows(
+	env string, envId, endpointId icingadbTypes.Binary,
+	selectCache func(dest interface{}, query string, args ...interface{}), _ *sqlx.Tx, idoRows []flappingRow,
+) (icingaDbUpdates, icingaDbInserts [][]interface{}, checkpoint interface{}) {
+	if len(idoRows) < 1 {
+		return
+	}
+
+	var cached []struct {
+		HistoryId     uint64
+		EventTime     int64
+		EventTimeUsec uint32
+	}
+	selectCache(
+		&cached, "SELECT history_id, event_time, event_time_usec FROM end_start_time WHERE history_id BETWEEN ? AND ?",
+		idoRows[0].FlappinghistoryId, idoRows[len(idoRows)-1].FlappinghistoryId,
+	)
+
+	// Needed for start time (see below).
+	cachedById := make(map[uint64]icingadbTypes.UnixMilli, len(cached))
+	for _, c := range cached {
+		cachedById[c.HistoryId] = convertTime(c.EventTime, c.EventTimeUsec)
+	}
+
+	var flappingHistory, flappingHistoryUpdates, allHistory []interface{}
+	for _, row := range idoRows {
+		ts := convertTime(row.EventTime, row.EventTimeUsec)
+
+		// Needed for ID (see below).
+		var start icingadbTypes.UnixMilli
+		if row.EventType == 1001 { // end
+			var ok bool
+			start, ok = cachedById[row.FlappinghistoryId]
+
+			if !ok {
+				continue
+			}
+		} else {
+			start = ts
+		}
+
+		name := row.Name1
+		if row.Name2 != "" {
+			name += "!" + row.Name2
+		}
+
+		typ := objectTypes[row.ObjecttypeId]
+		hostId := calcObjectId(env, row.Name1)
+		serviceId := calcServiceId(env, row.Name1, row.Name2)
+		startTime := float64(start.Time().UnixMilli())
+		flappingHistoryId := hashAny([]interface{}{env, name, startTime})
+
+		if row.EventType == 1001 { // end
+			// The start counterpart should already have been inserted.
+			flappingHistoryUpdates = append(flappingHistoryUpdates, &FlappingEnd{
+				flappingHistoryId, ts, icingadbTypes.Float{NullFloat64: row.PercentStateChange},
+			})
```

Comment:
But do you have to? Why not just perform the same operation that would also be used when writing fresh history?

Reply:
You mean INSERT-ignore? /s ;-)

Reply:
Do you actually consider it a better idea to send 10+ columns per row than 3?

Reply:
> You mean INSERT-ignore? /s ;-)

I don't think so or I don't understand what you're trying to say here. In the regular history sync, every event goes through this upsert, even those that would typically just update a previously inserted row: https://github.com/Icinga/icingadb/blob/0061888c404f87158d449d9b3ced82fca269a13b/pkg/icingadb/history/sync.go#L256

> Do you actually consider it a better idea to send 10+ columns per row than 3?

There probably is no "this is strictly better in every aspect solution", I'm trying to understand the reasons why it's done this way here.

The basic idea is that the faster execution of batched upserts outweighs the cost of sending some extra values, which it should given that this unlikely do be limited by network throughput. But given that the flapping history should be small compared to other types, this won't make a huge difference overall I guess (things would look differently if we were talking about state history here).

Using upserts could give some more benefits:
1. If you've lost the start event in the IDO history, there would be no row to update. When using upserts, you could at least migrate the flapping end event.
2. Removes all the extra code needed to handle updates just for this single case.

But maybe there's some reason why this wouldn't work in the first place (maybe something like using the upsert would overwrite valid data with zero values because things are no longer known in the end event or something like that), that's why I'm asking, because if that's the case, we don't need to have this discussion.